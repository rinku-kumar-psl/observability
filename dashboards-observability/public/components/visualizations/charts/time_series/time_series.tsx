/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { isEmpty, take } from 'lodash';
import { Plt } from '../../plotly/plot';
import { LONG_CHART_COLOR, PLOTLY_COLOR } from '../../../../../common/constants/shared';

export const TimeSeries = ({ visualizations, layout, config }: any) => {
  console.log("visualizations:::>", visualizations)
  const { vis } = visualizations;
  const {
    data,
    metadata: { fields },
  } = visualizations.data.rawVizData;
  const { isUniColor } = vis?.visConfig;
  const lastIndex = fields.length - 1;
  const { dataConfig = {}, layoutConfig = {} } = visualizations?.data?.userConfigs;
  const xaxis =
    dataConfig.valueOptions && dataConfig.valueOptions.xaxis ? dataConfig.valueOptions.xaxis : [];
  const yaxis =
    dataConfig.valueOptions && dataConfig.valueOptions.xaxis ? dataConfig?.valueOptions.yaxis : [];
  const seriesOrientation =
    dataConfig?.chartOptions?.orientation &&
    dataConfig.chartOptions.orientation[0] &&
    dataConfig.chartOptions.orientation[0].orientationId
      ? dataConfig.chartOptions.orientation[0].orientationId
      : visualizations.vis.orientation;
  const { defaultAxes } = visualizations.data;

  const isVertical = seriesOrientation === 'v';

  // when: stackLength = 1 and length of result buckets < 16 and chart is not unicolor
  // Else each stacked series has its own color using colorway
  let marker = {};
  if (lastIndex === 1 && data[fields[lastIndex].name].length < 16 && !isUniColor) {
    marker = {
      color: data[fields[lastIndex].name].map((_: string, index: number) => {
        return PLOTLY_COLOR[index % PLOTLY_COLOR.length];
      }),
    };
  }

  let valueSeries;
  if (!isEmpty(xaxis) && !isEmpty(yaxis)) {
    valueSeries = isVertical ? [...yaxis] : [...xaxis];
  } else {
    valueSeries = defaultAxes.yaxis || take(fields, lastIndex > 0 ? lastIndex : 1);
  }

  const timeSeries = valueSeries.map((field: any) => {
    return {
      x: isVertical
        ? data[!isEmpty(xaxis) ? xaxis[0].label : fields[lastIndex].name]
        : data[field.name],
      y: isVertical
        ? data[field.name]
        : data[!isEmpty(yaxis) ? yaxis[0]?.label : fields[lastIndex].name],
      type: vis.type,
      marker,
      name: field.name,
      orientation: seriesOrientation,
    };
  });
  console.log("fields::", fields);
  console.log("valueSeries:", valueSeries);

  console.log("timeSeries::", timeSeries);
  // If chart has length of result buckets < 16
  // then use the LONG_CHART_COLOR for all the series in the chart
  const plotlyColorway =
    data[fields[lastIndex].name].length < 16 ? PLOTLY_COLOR : [LONG_CHART_COLOR];

  const mergedLayout = {
    colorway: plotlyColorway,
    ...layout,
    ...(layoutConfig.layout && layoutConfig.layout),
    title: dataConfig?.panelOptions?.title || layoutConfig.layout?.title || '',
    barmode:
      dataConfig?.chartOptions?.mode &&
      dataConfig.chartOptions.mode[0] &&
      dataConfig.chartOptions.mode[0].modeId
        ? dataConfig.chartOptions.mode[0].modeId
        : '',
  };

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };

  return <Plt data={timeSeries} layout={mergedLayout} config={mergedConfigs} />;
};
