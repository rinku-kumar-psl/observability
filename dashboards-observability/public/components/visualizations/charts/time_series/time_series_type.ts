/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { TIME_SERIES } from '../../../../../common/constants/explorer';
import { ConfigValueOptions } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import { LensIconChartBar } from '../../assets/chart_bar';
import { getPlotlyCategory, getPlotlySharedConfigs } from '../shared/shared_configs';
import { TimeSeries } from './time_series';


const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();

export const createTimeSeriesTypeDefinition = (params: any) => ({
  name: TIME_SERIES,
  type: 'scatter',
  id: TIME_SERIES,
  label: 'Time Series',
  fullLabel: 'Time Series',
  iconType: 'visBarVerticalStacked',
  selection: {
    dataLoss: 'nothing',
  },
  category: VIS_CATEGORY.BASICS,
  icon: LensIconChartBar,
  categoryAxis: 'xaxis',
  seriesAxis: 'yaxis',
  orientation: 'v',
  component: TimeSeries,
  editorConfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Data',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          {
            id: 'value_options',
            name: 'Value options',
            editor: ConfigValueOptions,
            mapTo: 'valueOptions',
            schemas: [
              {
                name: 'X-axis',
                isSingleSelection: true,
                component: null,
                mapTo: 'xaxis',
              },
              {
                name: 'Y-axis',
                isSingleSelection: false,
                component: null,
                mapTo: 'yaxis',
              },
            ],
          },
        ],
      },
      {
        id: 'style-panel',
        name: 'Layout',
        mapTo: 'layoutConfig',
        editor: ConfigEditor,
        content: [],
      },
    ],
  },
  visConfig: {
    layout: {
      ...sharedConfigs.layout,
    },
    config: {
      ...sharedConfigs.config,
    },
    isUniColor: false,
  },
});
