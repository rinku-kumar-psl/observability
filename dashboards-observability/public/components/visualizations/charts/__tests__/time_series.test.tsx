/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { SAMPLE_VISUALIZATIONS_VIS_TIME_SERIES } from '../../../../../test/constants';
import {
  LAYOUT_CONFIG,
  SAMPLE_VISUALIZATIONS
} from '../../../../../test/event_analytics_constants';
import { TimeSeries } from '../time_series/time_series';

describe('Time-Series component', () => {
  configure({ adapter: new Adapter() });

  const COMPLETE_VISUALIZATIONS = {
    ...SAMPLE_VISUALIZATIONS,
    vis: SAMPLE_VISUALIZATIONS_VIS_TIME_SERIES,
  };
  it('Renders Time-Series component', async () => {
    const wrapper = mount(
      <TimeSeries
        visualizations={COMPLETE_VISUALIZATIONS}
        name="Time series"
        layoutConfig={LAYOUT_CONFIG}
        isUniColor={true}
      />
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
