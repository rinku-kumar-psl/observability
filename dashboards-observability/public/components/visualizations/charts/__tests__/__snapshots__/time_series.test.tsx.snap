// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time-Series component Renders Time-Series component 1`] = `
<TimeSeries
  isUniColor={true}
  layoutConfig={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  name="Time series"
  visualizations={
    Object {
      "data": Object {
        "count()": Array [
          2549,
          9337,
          1173,
        ],
        "defaultAxes": Object {
          "xaxis": Array [
            Object {
              "name": "new_timestamp",
              "type": "timestamp",
            },
          ],
          "yaxis": Array [
            Object {
              "name": "avg(AvgTicketPrice)",
              "type": "double",
            },
          ],
        },
        "rawVizData": Object {
          "data": Object {
            "avg(AvgTicketPrice)": Array [
              574.6643260854536,
              575.7916175023022,
              599.2321691707689,
              594.5192950694553,
            ],
            "new_timestamp": Array [
              "2022-04-04 00:00:00",
              "2022-04-05 00:00:00",
              "2022-04-06 00:00:00",
              "2022-04-07 00:00:00",
            ],
          },
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "avg(AvgTicketPrice)",
                "type": "double",
              },
              Object {
                "name": "new_timestamp",
                "type": "timestamp",
              },
            ],
          },
        },
        "span(timestamp,1M)": Array [
          "2021-05-01 00:00:00",
          "2021-06-01 00:00:00",
          "2021-07-01 00:00:00",
        ],
        "userConfigs": Object {
          "dataConfig": Object {},
          "layoutConfig": Object {},
        },
      },
      "jsonData": Array [
        Object {
          "count()": 2549,
          "span(timestamp,1M)": "2021-05-01 00:00:00",
        },
        Object {
          "count()": 9337,
          "span(timestamp,1M)": "2021-06-01 00:00:00",
        },
        Object {
          "count()": 2549,
          "span(timestamp,1M)": "2021-07-01 00:00:00",
        },
      ],
      "metadata": Object {
        "fields": Array [
          Object {
            "name": "count()",
            "type": "integer",
          },
          Object {
            "name": "span(timestamp,1M)",
            "type": "timestamp",
          },
        ],
      },
      "vis": Object {
        "category": "Visualizations",
        "categoryAxis": "xaxis",
        "editorConfig": Object {
          "panelTabs": Array [
            Object {
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Data",
              "sections": Array [
                Object {
                  "id": "value_options",
                  "mapTo": "valueOptions",
                  "name": "Value options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "xaxis",
                      "name": "X-axis",
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "yaxis",
                      "name": "Y-axis",
                    },
                  ],
                },
                Object {
                  "id": "chart_options",
                  "mapTo": "chartOptions",
                  "name": "Chart options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "orientation",
                      "name": "Orientation",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                          Object {
                            "name": "Horizontal",
                            "orientationId": "h",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "mode",
                      "name": "Mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                          Object {
                            "modeId": "stack",
                            "name": "Stack",
                          },
                        ],
                      },
                    },
                  ],
                },
              ],
            },
            Object {
              "content": Array [],
              "id": "style-panel",
              "mapTo": "layoutConfig",
              "name": "Layout",
            },
          ],
        },
        "fullLabel": "Time Series",
        "iconType": "visBarVerticalStacked",
        "id": "time_series",
        "label": "Time Series",
        "name": "time_series",
        "orientation": "v",
        "selection": Object {
          "dataLoss": "nothing",
        },
        "seriesAxis": "yaxis",
        "type": "scatter",
        "visConfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "isUniColor": false,
          "layout": Object {
            "height": 500,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "showlegend": true,
          },
        },
      },
    }
  }
>
  <Plt
    config={Object {}}
    data={
      Array [
        Object {
          "mode": "line",
          "name": "avg(AvgTicketPrice)",
          "type": "scatter",
          "x": Array [
            "2022-04-04 00:00:00",
            "2022-04-05 00:00:00",
            "2022-04-06 00:00:00",
            "2022-04-07 00:00:00",
          ],
          "y": Array [
            574.6643260854536,
            575.7916175023022,
            599.2321691707689,
            594.5192950694553,
          ],
        },
      ]
    }
    layout={
      Object {
        "title": "",
      }
    }
  >
    <PlotlyComponent
      config={
        Object {
          "displayModeBar": false,
        }
      }
      data={
        Array [
          Object {
            "mode": "line",
            "name": "avg(AvgTicketPrice)",
            "type": "scatter",
            "x": Array [
              "2022-04-04 00:00:00",
              "2022-04-05 00:00:00",
              "2022-04-06 00:00:00",
              "2022-04-07 00:00:00",
            ],
            "y": Array [
              574.6643260854536,
              575.7916175023022,
              599.2321691707689,
              594.5192950694553,
            ],
          },
        ]
      }
      debug={false}
      layout={
        Object {
          "autosize": true,
          "barmode": "stack",
          "hovermode": "closest",
          "legend": Object {
            "orientation": "h",
            "traceorder": "normal",
          },
          "margin": Object {
            "b": 30,
            "l": 30,
            "pad": 4,
            "r": 5,
            "t": 5,
          },
          "showlegend": false,
          "title": "",
          "xaxis": Object {
            "automargin": true,
            "rangemode": "normal",
            "showgrid": true,
            "zeroline": false,
          },
          "yaxis": Object {
            "rangemode": "normal",
            "showgrid": true,
            "zeroline": false,
          },
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useResizeHandler={true}
    >
      <div
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </PlotlyComponent>
  </Plt>
</TimeSeries>
`;
